src/allmydata/scripts/tahoe_backup.py

```diff
@@ -313,14 +263,353 @@ def upload(self, childpath):
+        self._directories_skipped = directories_skipped
+        self.dircap = dircap
+
+
+    def any_skips(self):
+        return self._files_skipped or self._directories_skipped
+
+
+    def report(self, verbosity, files_checked, directories_checked):
+        result = []
+
+        if verbosity >= 1:
+            result.append(
+                " %d files uploaded (%d reused),"
+                " %d files skipped,"
+                " %d directories created (%d reused),"
+                " %d directories skipped" % (
+                    self._files_created,
+                    self._files_reused,
+                    self._files_skipped,
+                    self._directories_created,
+                    self._directories_reused,
+                    self._directories_skipped,
+                ),
+            )
+
+        if verbosity >= 2:
+            result.append(
+                " %d files checked, %d directories checked" % (
+                    files_checked,
+                    directories_checked,
+                ),
+            )
+        # calc elapsed time, omitting microseconds
+        elapsed_time = str(
+            self._end_timestamp - self._start_timestamp
+        ).split('.')[0]
+        result.append(" backup done, elapsed time: %s" % (elapsed_time,))
+
+        return "\n".join(result)
+
+
+
+class BackupProgress(object):
+    # Would be nice if this data structure were immutable and its methods were
+    # transformations that created a new slightly different object.  Not there
+    # yet, though.
+    def __init__(self, warn, start_timestamp, target_count):
+        self._warn = warn
+        self._start_timestamp = start_timestamp
+        self._target_count = target_count
+        self._files_created = 0
+        self._files_reused = 0
+        self._files_skipped = 0
+        self._directories_created = 0
+        self._directories_reused = 0
+        self._directories_skipped = 0
+        self.last_dircap = None
+        self._create_contents = {}
+        self._compare_contents = {}
+
+
+    def report(self, now):
+        report_format = (
+            "Backing up {target_progress}/{target_total}... {elapsed} elapsed..."
+        )
+        return report_format.format(
+            target_progress=(
+                self._files_created
+                + self._files_reused
+                + self._files_skipped
+                + self._directories_created
+                + self._directories_reused
+                + self._directories_skipped
+            ),
+            target_total=self._target_count,
+            elapsed=self._format_elapsed(now - self._start_timestamp),
+        )
+
+
+    def _format_elapsed(self, elapsed):
+        seconds = elapsed.total_seconds()
+        hours = int(seconds / 3600)
+        minutes = int(seconds / 60 % 60)
+        seconds = int(seconds % 60)
+        return "{}h {}m {}s".format(
+            hours,
+            minutes,
+            seconds,
+        )
```

Comment:
There is `allmydata.util.time_format.format_delta` to consider as well here.

(There are several spots in tahoe code that could probably just use "humanize" instead but I don't think we currently depend on that library.)

Reply:
Cool, didn't know about that.  Looks like since I have datetimes instead of ints, though, it's more trouble to use than it's worth.  `format_delta` wants start and stop ints and converting a datetime to a posix timestamp (_still_!?) involves a ton of wrangling, I think.

+1 to adding a humanize dependency though.